### PR TITLE
fix: Fix TE ownership validation in Tracker importer [DHIS2-17251]

### PR DESCRIPTION
**Pull request by enricocolasante**
In the importer we need to validate access to Tracked Entity the same way we are now doing in the exporter.
We have 3 different cases in the importer:

- CREATE: Check capture scope of the user using registration unit
- UPDATE or DELETE: Check if the user has access to any TE/program pair, by first matching by the owner, or the registering org unit if no owner is found

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/validation/validator/trackedentity/SecurityOwnershipValidator.java [hunk 1]
```diff
@@ -91,14 +95,18 @@ public void validate(
                 .getOrganisationUnit()
             : bundle.getPreheat().getOrganisationUnit(trackedEntity.getOrgUnit());
 
-    // If trackedEntity is newly created, or going to be deleted, capture
-    // scope has to be checked
-    if (strategy.isCreate() || strategy.isDelete()) {
+    // If trackedEntity is newly created, capture scope has to be checked
```
Comment (muilpp) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/validation/validator/trackedentity/SecurityOwnershipValidator.java:98]: can we either remove the comment or add that we also validate write access to the tracked entity type?

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/validation/validator/trackedentity/SecurityOwnershipValidator.java [hunk 2]
```diff
@@ -91,14 +95,18 @@ public void validate(
...
+    if (strategy.isCreate()) {
+      checkTeiTypeWriteAccess(reporter, bundle, trackedEntity, trackedEntityType);
```
Comment (muilpp) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/validation/validator/trackedentity/SecurityOwnershipValidator.java:100]: rename the method to check**TeT**ypeWriteAccess

**Review by muilpp — COMMENTED** [review]
lgtm, just a couple of details I've seen.

I might add a test or two where a TE has been transferred, to validate a user with access to the transferred org unit can delete that TE, and a user with access to the registration org unit only, can't.

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/validation/validator/trackedentity/SecurityOwnershipValidator.java [hunk 3]
```diff
@@ -91,14 +95,18 @@ public void validate(
...
       checkOrgUnitInCaptureScope(reporter, bundle, trackedEntity, organisationUnit);
     }
-    // if its to update trackedEntity, search scope has to be checked
+    // if its to update or delete trackedEntity, write access has to be checked
```
Comment (muilpp) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/validation/validator/trackedentity/SecurityOwnershipValidator.java:103]: similar to the one above, I think we either expand on why write access has to be checked, or we remove the comment altogether.

**Comment by enricocolasante** [comment]
> lgtm, just a couple of details I've seen.
> 
> I might add a test or two where a TE has been transferred, to validate a user with access to the transferred org unit can delete that TE, and a user with access to the registration org unit only, can't.

Added integration tests to check that update and delete for transferred TE use ownership and not the registration org unit.
Create was already covered by existing tests

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/validation/validator/trackedentity/SecurityOwnershipValidator.java [hunk 4]
```diff
@@ -145,15 +153,25 @@ private void checkOrgUnitInCaptureScope(
...
+            .toList();
+    if (!programsToCheck.isEmpty()
+        && programsToCheck.stream()
+            .noneMatch(
+                p ->
+                    trackerAccessManager
+                        .canWrite(UserDetails.fromUser(bundle.getUser()), te, p, false)
+                        .isEmpty())) {
+      reporter.addError(dto, ValidationCode.E1003, te.getUid());
     }
```
Comment (lucaCambi77) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/validation/validator/trackedentity/SecurityOwnershipValidator.java]: a question about this. Should this be a general rule for checking access to write a te? Could this be moved to the ACL directly as we did for `canRead` of a te?

**Review by muilpp — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/validation/validator/trackedentity/SecurityOwnershipValidator.java (hunk 4 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/validation/validator/trackedentity/SecurityOwnershipValidator.java]: This is an outdated version of the code.
I already moved it to the ACL

**Review by lucaCambi77 — APPROVED** [review]
you moved it in another PR? not sure what you mean but i don't want to block

**Pull request merged**